### PR TITLE
Uses MCMCSpec in distregbuilder

### LINKED ISSUE #273: Improve `MCMCSpec` Jitter Mechanism for Greater Flexibility and Compatibility
The current implementation of `jitter` in `MCMCSpec` is limiting due to two key assumptions:

1. The jitter distribution produces a scalar draw.
2. The jitter is added to the existing value.

These constraints have prevent to migrate  `distreg` to use `MCMCSpec`. Specifically,

* Scalar-only jitter does not accommodate variable shapes. It might be useful and necessary to have initialization noise that matches the shape of the variable -- either through matching the event shape or batch shape of the variable. Even if we do not want to allow it, we should provide better error messages, by testing that the batch and event shape are `()`.
* Addition-only jitter is inflexible. In some cases, rather than adding to a provided value, it is preferable to replace the value entirely using a sample from a distribution (as we do in the distreg builder).

### Proposal

1. Support shape-aware jitter distributions
    Allow users to provide a jitter distribution whose shape matches the shape of the variable (either via batch shape or event shape or a combination of those)

2. Support replacing the value instead of always adding
   Introduce a mechanism (e.g., a flag) that allows users to specify whether the jitter should be added, multiplied or replace the original value.

### PR DESCRIPTION
**Pull request by wiep**
Uses the new interface in distreg builder. Also, improves the type specification in `MCMCSpec`.

## Functional Changes
- ~beta values are additively jittered instead of initialized uniformly between -2 and 2. Old behavior is currently not possible. See #273.~

## Todo

- [x] Resolve #273 and restore old behavior
- [x] Raise error or warning if variables without MCMCSpec are encountered
- [x] Add tests for sampling shapes and jitter types (additive, multiplicative, replacement, none)
- [x] Agree on name for `KernelFactory`
- [x] Agree on a way how to handle different jitter strategies
- [x] Set jitter strategy to additive for beta and replacement for tau

## Optional

- Remove groups from builder

### CODE REVIEW

**Comment by jobrachem** [comment]
See my comments in https://github.com/liesel-devs/liesel/issues/273

Also, I am confused about what you say about the old behavior of jittering for beta values. To me it seems, that it was indeed additive:

```python
def _beta_jitter_fn(key, val):
    return val + jax.random.uniform(key, val.shape, val.dtype, -2.0, 2.0)
```

Only the jittering for the variances seems to have been non-additive:

```python
def _tau2_jitter_fn(key, val):
    return jax.random.truncated_normal(key, 0.0, 1e2, val.shape, val.dtype)
```

Regarding the updated jitter functionality: We might want to consider just changing the API slightly to allow user-specified functions, see 3.) in https://github.com/liesel-devs/liesel/issues/273#issuecomment-2853387283 . That would reduce the need to anticipate and classify different types of jittering with an Enum.

**Comment by wiep** [comment]
You are right. Tau is not additive. my bad

File: src/liesel/model/distreg.py [hunk 1]
```diff
@@ -294,36 +321,28 @@ def dist_reg_mcmc(
...
+        name for kernel in builder.kernels for name in kernel.position_keys
+    ]
 
-    builder.set_jitter_fns(jitter_fns)
+    missing_vars = set(var_names) - set(var_names_in_kernels)
+    if missing_vars:
+        raise RuntimeError(
+            f"The following parameters are not associated with any kernel: "
+            f"{missing_vars}"
+        )
```
Comment (jobrachem) [inline src/liesel/model/distreg.py:340]: It seems like this would prevent users from removing inference info from a parameter. However, that may sometimes be very helpful for debugging. I'm not sure that enforcing a kernel for all parameters is necessary.

File: src/liesel/goose/mcmc_spec.py [hunk 2]
```diff
@@ -217,12 +250,19 @@ class MCMCSpec:
...
+        - `JitterType.MULTIPLICATIVE`: Multiplicative jitter is applied.
+        - `JitterType.REPLACEMENT`: Value is replaced when jitter is applied.
     """
 
-    kernel: Callable[..., Kernel]
+    kernel: KernelFactory
     kernel_kwargs: dict[str, Any] = field(default_factory=dict)
     kernel_group: str | None = None
     jitter_dist: tfd.Distribution | None = None
+    jitter_method: JitterMethod = JitterMethod.ADDITIVE
```
[the remark this answers is not in the document]
Comment (jobrachem) [reply src/liesel/goose/mcmc_spec.py:265]: this is not a critical issue, though. I think the enum is totally fine. I just added it to goose's exports, so that `JitterMethod` does not always have to be imported manually.

**Review by jobrachem — COMMENTED** [review]
- I like the changes, have only a few comments, specifically about `JitterMethods.NONE` and the `RuntimeError` in `dist_reg_mcmc`.
- I like the `KernelFactory`, including the name.

File: src/liesel/model/distreg.py (hunk 1 above)
Comment (jobrachem) [reply src/liesel/model/distreg.py:340]: This is a minor point, because it would be trivial for users to simply call `LieselMCMC` directly.

File: src/liesel/goose/mcmc_spec.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wiep) [reply src/liesel/goose/mcmc_spec.py:265]: both versions are fine for me. I just did not know about the Literal type.

File: src/liesel/goose/mcmc_spec.py [hunk 3]
```diff
@@ -243,12 +283,37 @@ def apply_jitter(self, seed: KeyArray, value: Array) -> Array:
         -------
         The jittered value with the same shape as the input.
         """
-        if self.jitter_dist is None:
+        if self.jitter_dist is None or self.jitter_method == JitterMethod.NONE:
```
[the remark this answers is not in the document]
Comment (wiep) [reply src/liesel/goose/mcmc_spec.py:286]: makes sense.

File: src/liesel/goose/mcmc_spec.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wiep) [reply src/liesel/goose/mcmc_spec.py:265]: I added a `__post_init__` as an early type check to MCMCSpec.

File: src/liesel/model/distreg.py (hunk 1 above)
Comment (wiep) [reply src/liesel/model/distreg.py:342]: I added that explicitly because of RLiesel -- however, I converted it into a warning, that is good enough and doesn't prevent the debugging.

File: src/liesel/goose/mcmc_spec.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jobrachem) [reply src/liesel/goose/mcmc_spec.py]: nice!

File: src/liesel/model/distreg.py (hunk 1 above)
Comment (jobrachem) [reply src/liesel/model/distreg.py:342]: sounds good to me!

**Review by jobrachem — APPROVED** [review]
